tests/solana_tests/optimizations.rs

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    borsh_encoding::{visit_mut, VisitorMut},
+    BorshToken, VirtualMachineBuilder,
+};
+use anchor_syn::idl::IdlInstruction;
+use once_cell::sync::Lazy;
+use serde::Deserialize;
+use solang::codegen::Options;
+use std::{
+    env::var,
+    fs::{read_dir, read_to_string, File},
+    io::BufReader,
+    path::{Path, PathBuf},
+};
+
+#[derive(Debug, Deserialize)]
+struct Calls {
+    constructor: Vec<BorshToken>,
+    function: Vec<(String, Vec<BorshToken>)>,
+}
+
+static NO_OPTIMIZATIONS: Lazy<Options> = Lazy::new(|| Options {
+    dead_storage: false,
+    constant_folding: false,
+    strength_reduce: false,
+    vector_to_slice: false,
+    common_subexpression_elimination: false,
+    ..Default::default()
+});
+
+#[test]
+fn optimizations() {
+    let calls = Path::new("tests/optimization_testcases/calls");
+
+    if let Ok(testname) = var("TESTNAME") {
+        run_tests(std::iter::once(calls.join(testname).with_extension("json")));
+    } else {
+        run_tests(read_dir(calls).unwrap().map(|entry| entry.unwrap().path()));
+    }
+}
+
+fn run_tests(iter: impl Iterator<Item = PathBuf>) {
+    for path in iter {
+        let file_stem = path.file_stem().unwrap();
+
+        // Known problematic test.
+        if file_stem == "a467d917a5158b613f94954d4a558a5cb61e6773" {
+            continue;
+        }
+
+        dbg!(file_stem);
+
+        let file = File::open(&path).unwrap();
+        let reader = BufReader::new(file);
+        let calls: Calls = serde_json::from_reader(reader).unwrap();
+
+        let path = Path::new("tests/optimization_testcases/programs")
+            .join(file_stem)
+            .with_extension("sol");
+        let program = read_to_string(path).unwrap();
+
+        run_one_test(
+            &program,
+            &calls,
+            [Options::default(), NO_OPTIMIZATIONS.clone()],
+        );
+    }
+}
+
+fn run_one_test(program: &str, calls: &Calls, opts: impl IntoIterator<Item = Options>) {
+    let mut results_prev: Option<Vec<Result<Option<BorshToken>, u64>>> = None;
+
+    for (i, opts) in opts.into_iter().enumerate() {
+        dbg!(i);
+
+        let mut results_curr = Vec::new();
+
+        let mut vm = VirtualMachineBuilder::new(program).opts(opts).build();
+
+        let data_account = vm.initialize_data_account();
+
+        results_curr.push(
+            vm.function("new")
+                .arguments(&calls.constructor)
+                .accounts(vec![("dataAccount", data_account)])
+                .call_with_error_code(),
+        );
+
+        for (name, args) in &calls.function {
+            if let Some(idl) = &vm.stack[0].idl {
+                let mut idl = idl.clone();
+
+                idl.instructions.push(IdlInstruction {
+                    name: name.clone(),
+                    docs: None,
+                    accounts: vec![],
+                    args: vec![],
+                    returns: None,
+                });
```

Comment:
The IDL is generated when you call `build`. Why did you need to push the instructions again here?

Reply:
To be honest, I don't really understand what is going on there.

I was imitating this example: https://github.com/hyperledger/solang/blob/373cff71f27515cba98f16c2f7632c2092c95041/tests/solana_tests/balance.rs#L384-L396

If you comment out that `if` block you cited in `optimizations.rs` and run:
```
TESTNAME=c3c446fbffb01f22ace1e7311385624ff08b40ae cargo test --test solana optimizations
```
you get `panicked at 'Function not found'`.

For convenience, here is [c3c446fbffb01f22ace1e7311385624ff08b40ae.sol](https://github.com/hyperledger/solang/blob/b1b2ac7d93a8c7b251e98e8a3ed6324be1d1516b/tests/optimization_testcases/programs/c3c446fbffb01f22ace1e7311385624ff08b40ae.sol).

Reply:
To be sure I understand the changes requested: remove the `if let Some(idl) ...` block and all testcases that produce "Function not found" errors?

Reply:
> all testcases that produce "Function not found" errors

It looks like  `c3c446fbffb01f22ace1e7311385624ff08b40ae` was the only one.

Reply:
> To be sure I understand the changes requested: remove the `if let Some(idl) ...` block and all testcases that produce "Function not found" errors?

Yes, exactly. `c3c446fbffb01f22ace1e7311385624ff08b40ae` must be the only case in which this error happens.